File: bindings/javascript/packages/core/src/ca-peps.ts
/**
 * Clifford-Assisted PEPS (2D) binding (C-side since v0.2.1, JS
 * binding since v0.4.12).
 *
 * Wraps ``src/algorithms/tensor_network/ca_peps.{c,h}``, the 2D
 * generalisation of CA-MPS.  Same Clifford-tableau + physical-MPS
 * split: Clifford gates are free (O(n) tableau bit-ops), and
 * non-Clifford gates conjugate through to Pauli rotations on the
 * physical factor.  Mirrors Python ``moonlab.ca_peps.CAPEPS`` and
 * Rust ``moonlab::ca_peps::CaPeps``.
 *
 * @example
 * ```typescript
 * import { CaPeps, PauliCode } from '@moonlab/quantum-core';
 *
 * // 2x3 lattice with bond-dim cap chi = 8.
 * const state = await CaPeps.create(2, 3, 8);
 * state.h(0).cnot(0, 1);
 * const z0 = state.expectPauliSingle(0, PauliCode.Z);
 * console.log(`<Z_0> = ${z0.toFixed(6)}`);
 * state.dispose();
 * ```
 */

import { getModule } from './wasm-loader';
import { PauliCode } from './mpdo';

// Re-export so `import { PauliCode } from '@moonlab/quantum-core'` keeps
// working when callers reach for CaPeps observables.
export { PauliCode } from './mpdo';

type CaPepsModule = {
  _moonlab_ca_peps_create: (lx: number, ly: number, chi: number) => number;
  _moonlab_ca_peps_free: (h: number) => void;
  _moonlab_ca_peps_clone: (h: number) => number;
  _moonlab_ca_peps_lx: (h: number) => number;
  _moonlab_ca_peps_ly: (h: number) => number;
  _moonlab_ca_peps_num_qubits: (h: number) => number;
  _moonlab_ca_peps_max_bond_dim: (h: number) => number;
  _moonlab_ca_peps_current_bond_dim: (h: number) => number;
  _moonlab_ca_peps_max_half_cut_entropy: (h: number) => number;
  _moonlab_ca_peps_norm: (h: number) => number;
  _moonlab_ca_peps_normalize: (h: number) => number;
  _moonlab_ca_peps_h: (h: number, q: number) => number;
  _moonlab_ca_peps_s: (h: number, q: number) => number;
  _moonlab_ca_peps_sdag: (h: number, q: number) => number;
  _moonlab_ca_peps_x: (h: number, q: number) => number;
  _moonlab_ca_peps_y: (h: number, q: number) => number;
  _moonlab_ca_peps_z: (h: number, q: number) => number;
  _moonlab_ca_peps_cnot: (h: number, c: number, t: number) => number;
  _moonlab_ca_peps_cz: (h: number, a: number, b: number) => number;
  _moonlab_ca_peps_rx: (h: number, q: number, theta: number) => number;
  _moonlab_ca_peps_ry: (h: number, q: number, theta: number) => number;
  _moonlab_ca_peps_rz: (h: number, q: number, theta: number) => number;
  _moonlab_ca_peps_t_gate: (h: number, q: number) => number;
  _moonlab_ca_peps_t_dagger: (h: number, q: number) => number;
  _moonlab_ca_peps_phase: (h: number, q: number, theta: number) => number;
  _moonlab_ca_peps_prob_z: (h: number, q: number, outPtr: number) => number;
  _moonlab_ca_peps_expect_pauli: (
    h: number,
    pauliPtr: number,
    outPtr: number,
  ) => number;
  _malloc: (n: number) => number;
  _free: (p: number) => void;
  HEAPU8: Uint8Array;
  HEAPF64: Float64Array;
};

/**
 * Owned WASM handle to a CA-PEPS state.
 *
 * Linear-index addressing throughout: qubit `q = x + Lx * y`.
 * Two-qubit gates require adjacent linear indices; vertical
 * neighbours sit `Lx` apart in linear order.
 */
export class CaPeps {
  private handle: number;
  private mod: CaPepsModule;
  private freed = false;

  private constructor(handle: number, mod: CaPepsModule) {
    this.handle = handle;
    this.mod = mod;
  }

  /** Allocate a CA-PEPS on `Lx x Ly` with per-bond chi cap
   *  `chiBond`.  Initial state `|0...0>` with `D = I`. */
  static async create(lx: number, ly: number, chiBond: number): Promise<CaPeps> {
    if (lx < 1 || ly < 1) {
      throw new RangeError(`lx=${lx}, ly=${ly}: both must be >= 1`);
    }
    if (chiBond < 1) {
      throw new RangeError(`chiBond must be >= 1, got ${chiBond}`);
    }
    const mod = (await getModule()) as unknown as CaPepsModule;
    const h = mod._moonlab_ca_peps_create(lx, ly, chiBond);
    if (!h) throw new Error('moonlab_ca_peps_create returned NULL');
    return new CaPeps(h, mod);
  }

  /** Release the underlying WASM allocation. */
  dispose(): void {
    if (this.freed) return;
    this.mod._moonlab_ca_peps_free(this.handle);
    this.freed = true;
  }

  /** Deep copy. */
  clone(): CaPeps {
    this._guardLive();
    const h = this.mod._moonlab_ca_peps_clone(this.handle);
    if (!h) throw new Error('moonlab_ca_peps_clone returned NULL');
    return new CaPeps(h, this.mod);
  }

  get lx(): number { this._guardLive(); return this.mod._moonlab_ca_peps_lx(this.handle); }
  get ly(): number { this._guardLive(); return this.mod._moonlab_ca_peps_ly(this.handle); }
  get numQubits(): number {
    this._guardLive();
    return this.mod._moonlab_ca_peps_num_qubits(this.handle);
  }
  get maxBondDim(): number {
    this._guardLive();
    return this.mod._moonlab_ca_peps_max_bond_dim(this.handle);
  }
  get currentBondDim(): number {
    this._guardLive();
    return this.mod._moonlab_ca_peps_current_bond_dim(this.handle);
  }
  /** `<phi | phi>` of the physical factor; ~1 modulo truncation. */
  get norm(): number {
    this._guardLive();
    return this.mod._moonlab_ca_peps_norm(this.handle);
  }
  /** Maximum half-cut von Neumann entanglement entropy of `|phi>`
   *  across all bipartitions, in nats. */
  get maxHalfCutEntropy(): number {
    this._guardLive();
    return this.mod._moonlab_ca_peps_max_half_cut_entropy(this.handle);
  }

  normalize(): this {
    this._check(
      this.mod._moonlab_ca_peps_normalize(this.handle),
      'normalize',
    );
    return this;
  }

  // ---- Clifford gates ---------------------------------------------------
  h(q: number): this   { return this._call1('h', q); }
  s(q: number): this   { return this._call1('s', q); }
  sdag(q: number): this { return this._call1('sdag', q); }
  x(q: number): this   { return this._call1('x', q); }
  y(q: number): this   { return this._call1('y', q); }
  z(q: number): this   { return this._call1('z', q); }
  cnot(control: number, target: number): this {
    return this._call2('cnot', control, target);
  }
  cz(a: number, b: number): this { return this._call2('cz', a, b); }

  // ---- Non-Clifford gates ----------------------------------------------
  rx(q: number, theta: number): this { return this._call1p('rx', q, theta); }
  ry(q: number, theta: number): this { return this._call1p('ry', q, theta); }
  rz(q: number, theta: number): this { return this._call1p('rz', q, theta); }
  t(q: number): this    { return this._call1('t_gate', q); }
  tdg(q: number): this  { return this._call1('t_dagger', q); }
  phase(q: number, theta: number): this {
    return this._call1p('phase', q, theta);
  }

  // ---- Observables -----------------------------------------------------

  /** Marginal `P(Z_q = +1)`. */
  probZ(q: number): number {
    this._guardQubit(q);
    const out = this.mod._malloc(8);
    try {
      this._check(
        this.mod._moonlab_ca_peps_prob_z(this.handle, q, out),
        'prob_z',
      );
      return this.mod.HEAPF64[out >> 3];
    } finally {
      this.mod._free(out);
    }
  }

  /** `<psi | P | psi>` for an n-qubit Pauli string.  Entry per
   *  qubit is `PauliCode` (0=I, 1=X, 2=Y, 3=Z).  Length must equal
   *  `numQubits`.  Returns `[re, im]` of the complex expectation. */
  expectPauli(pauli: readonly PauliCode[] | Uint8Array): [number, number] {
    this._guardLive();
    const n = this.numQubits;
    if (pauli.length !== n) {
      throw new RangeError(
        `pauli.length=${pauli.length} != numQubits=${n}`,
      );
    }
    const pauliPtr = this.mod._malloc(n);
    const outPtr = this.mod._malloc(16);  // double _Complex = 2 doubles
    try {
      for (let i = 0; i < n; i++) {
        this.mod.HEAPU8[pauliPtr + i] = pauli[i];
      }
      this._check(
        this.mod._moonlab_ca_peps_expect_pauli(this.handle, pauliPtr, outPtr),
        'expect_pauli',
      );
      const idx = outPtr >> 3;
      return [this.mod.HEAPF64[idx], this.mod.HEAPF64[idx + 1]];
    } finally {
      this.mod._free(pauliPtr);
      this.mod._free(outPtr);
    }
  }

  /** `<psi | P_q | psi>` for a single-site Pauli on qubit `q`,
   *  identity elsewhere.  Returns the real part only (single-site
   *  Pauli expectations are real up to roundoff). */
  expectPauliSingle(q: number, pauli: PauliCode): number {
    this._guardQubit(q);
    const n = this.numQubits;
    const ops = new Uint8Array(n);  // zero-init = identity
    ops[q] = pauli;
    return this.expectPauli(ops)[0];
  }

  // ---- Internal helpers ------------------------------------------------

  private _guardLive(): void {
    if (this.freed) throw new Error('CaPeps is disposed');
  }

  private _guardQubit(q: number): void {
    this._guardLive();
    const n = this.numQubits;
    if (q < 0 || q >= n) {
      throw new RangeError(`qubit ${q} out of [0, ${n})`);
    }
  }

  private _call1(name: string, q: number): this {
    this._guardQubit(q);
    const m = this.mod as unknown as Record<string, (...args: number[]) => number>;
    this._check(m[`_moonlab_ca_peps_${name}`](this.handle, q), name);
    return this;
  }

  private _call1p(name: string, q: number, theta: number): this {
    this._guardQubit(q);
    const m = this.mod as unknown as Record<string, (...args: number[]) => number>;
    this._check(m[`_moonlab_ca_peps_${name}`](this.handle, q, theta), name);
    return this;
  }

  private _call2(name: string, a: number, b: number): this {
    this._guardQubit(a);
    this._guardQubit(b);
    const m = this.mod as unknown as Record<string, (...args: number[]) => number>;
    this._check(m[`_moonlab_ca_peps_${name}`](this.handle, a, b), name);
    return this;
  }

  private _check(rc: number, fn: string): void {
    if (rc !== 0) throw new Error(`moonlab_ca_peps_${fn} rc=${rc}`);
  }
}

File: bindings/rust/moonlab/src/ca_peps.rs
//! Clifford-Assisted PEPS (2D) -- safe Rust wrapper since v0.4.11.
//!
//! Wraps `src/algorithms/tensor_network/ca_peps.{c,h}`, the 2D
//! generalisation of CA-MPS.  Same Clifford-tableau + physical-MPS
//! split as [`crate::ca_mps::CaMps`]: Clifford gates are free
//! (O(n) tableau updates), and non-Clifford gates conjugate through
//! to Pauli rotations on the physical factor.  Mirrors the v0.2.1
//! Python `moonlab.ca_peps.CAPEPS` surface.
//!
//! ## Quick start
//!
//! ```no_run
//! use moonlab::ca_peps::{CaPeps, PauliCode};
//!
//! // 2x3 lattice with bond-dim cap chi = 8.
//! let mut state = CaPeps::new(2, 3, 8).unwrap();
//! state.h(0).unwrap();
//! state.cnot(0, 1).unwrap();
//! // Linear index = x + Lx*y; for Lx=2, qubit 2 is (0, 1).
//! let z0 = state.expect_pauli_single(0, PauliCode::Z).unwrap();
//! ```

use crate::error::{QuantumError, Result};
use moonlab_sys::{
    ca_peps_error_t, moonlab_ca_peps_clone, moonlab_ca_peps_cnot,
    moonlab_ca_peps_create, moonlab_ca_peps_current_bond_dim,
    moonlab_ca_peps_cz, moonlab_ca_peps_expect_pauli, moonlab_ca_peps_free,
    moonlab_ca_peps_h, moonlab_ca_peps_lx, moonlab_ca_peps_ly,
    moonlab_ca_peps_max_bond_dim, moonlab_ca_peps_max_half_cut_entropy,
    moonlab_ca_peps_norm, moonlab_ca_peps_normalize, moonlab_ca_peps_num_qubits,
    moonlab_ca_peps_phase, moonlab_ca_peps_prob_z, moonlab_ca_peps_rx,
    moonlab_ca_peps_ry, moonlab_ca_peps_rz, moonlab_ca_peps_s,
    moonlab_ca_peps_sdag, moonlab_ca_peps_t as moonlab_ca_peps_handle_t,
    moonlab_ca_peps_t_dagger, moonlab_ca_peps_t_gate, moonlab_ca_peps_x,
    moonlab_ca_peps_y, moonlab_ca_peps_z, __BindgenComplex,
};
use std::ptr;

/// Pauli code per qubit for [`CaPeps::expect_pauli`] /
/// [`CaPeps::expect_pauli_single`].  Matches the C header
/// convention `0=I, 1=X, 2=Y, 3=Z`.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
#[repr(u8)]
pub enum PauliCode {
    I = 0,
    X = 1,
    Y = 2,
    Z = 3,
}

/// Owned CA-PEPS handle.
pub struct CaPeps {
    ptr: *mut moonlab_ca_peps_handle_t,
}

impl CaPeps {
    /// Allocate a CA-PEPS on `Lx x Ly` qubits with per-bond
    /// bond-dim cap `chi_bond`.  Initial state is `|0...0>` with
    /// `D = I`.
    pub fn new(lx: u32, ly: u32, chi_bond: u32) -> Result<Self> {
        if lx == 0 || ly == 0 {
            return Err(QuantumError::Ffi(format!(
                "Lx={lx}, Ly={ly}: both dimensions must be >= 1"
            )));
        }
        if chi_bond == 0 {
            return Err(QuantumError::Ffi(
                "chi_bond must be >= 1".to_string(),
            ));
        }
        let ptr = unsafe { moonlab_ca_peps_create(lx, ly, chi_bond) };
        if ptr.is_null() {
            return Err(QuantumError::Ffi(
                "moonlab_ca_peps_create returned NULL".to_string(),
            ));
        }
        Ok(Self { ptr })
    }

    /// Lx, the lattice's x extent.
    pub fn lx(&self) -> u32 { unsafe { moonlab_ca_peps_lx(self.ptr) } }
    /// Ly, the lattice's y extent.
    pub fn ly(&self) -> u32 { unsafe { moonlab_ca_peps_ly(self.ptr) } }
    /// Total number of qubits `Lx * Ly`.
    pub fn num_qubits(&self) -> u32 {
        unsafe { moonlab_ca_peps_num_qubits(self.ptr) }
    }
    /// Configured bond-dim cap.
    pub fn max_bond_dim(&self) -> u32 {
        unsafe { moonlab_ca_peps_max_bond_dim(self.ptr) }
    }
    /// Current maximum bond dimension across all bonds.
    pub fn current_bond_dim(&self) -> u32 {
        unsafe { moonlab_ca_peps_current_bond_dim(self.ptr) }
    }
    /// `<phi | phi>` of the physical factor; should be 1 modulo
    /// truncation noise.
    pub fn norm(&self) -> f64 { unsafe { moonlab_ca_peps_norm(self.ptr) } }
    /// Maximum half-cut von Neumann entanglement entropy of `|phi>`
    /// across all bipartitions, in nats.
    pub fn max_half_cut_entropy(&self) -> f64 {
        unsafe { moonlab_ca_peps_max_half_cut_entropy(self.ptr) }
    }

    /// Normalise the physical factor in place.
    pub fn normalize(&mut self) -> Result<()> {
        check(unsafe { moonlab_ca_peps_normalize(self.ptr) })
    }

    // ---- Clifford gates (tableau-only; O(n) bit ops) -------------------

    pub fn h(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_h(self.ptr, q) })?;
        Ok(self)
    }
    pub fn s(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_s(self.ptr, q) })?;
        Ok(self)
    }
    pub fn sdag(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_sdag(self.ptr, q) })?;
        Ok(self)
    }
    pub fn x(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_x(self.ptr, q) })?;
        Ok(self)
    }
    pub fn y(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_y(self.ptr, q) })?;
        Ok(self)
    }
    pub fn z(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_z(self.ptr, q) })?;
        Ok(self)
    }
    pub fn cnot(&mut self, control: u32, target: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_cnot(self.ptr, control, target) })?;
        Ok(self)
    }
    pub fn cz(&mut self, a: u32, b: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_cz(self.ptr, a, b) })?;
        Ok(self)
    }

    // ---- Single-qubit non-Clifford rotations ---------------------------

    pub fn rx(&mut self, q: u32, theta: f64) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_rx(self.ptr, q, theta) })?;
        Ok(self)
    }
    pub fn ry(&mut self, q: u32, theta: f64) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_ry(self.ptr, q, theta) })?;
        Ok(self)
    }
    pub fn rz(&mut self, q: u32, theta: f64) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_rz(self.ptr, q, theta) })?;
        Ok(self)
    }
    pub fn t_gate(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_t_gate(self.ptr, q) })?;
        Ok(self)
    }
    pub fn t_dagger(&mut self, q: u32) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_t_dagger(self.ptr, q) })?;
        Ok(self)
    }
    pub fn phase(&mut self, q: u32, theta: f64) -> Result<&mut Self> {
        check(unsafe { moonlab_ca_peps_phase(self.ptr, q, theta) })?;
        Ok(self)
    }

    // ---- Observables ---------------------------------------------------

    /// `<psi | P | psi>` for an n-qubit Pauli string `pauli` of
    /// length `num_qubits()`.  Entry per qubit is `0=I, 1=X, 2=Y, 3=Z`.
    /// Returns the complex expectation value as `(real, imag)`.
    pub fn expect_pauli(&self, pauli: &[u8]) -> Result<(f64, f64)> {
        if pauli.len() != self.num_qubits() as usize {
            return Err(QuantumError::Ffi(format!(
                "pauli length {} != num_qubits {}",
                pauli.len(),
                self.num_qubits()
            )));
        }
        let mut out = __BindgenComplex { re: 0.0, im: 0.0 };
        check(unsafe {
            moonlab_ca_peps_expect_pauli(self.ptr, pauli.as_ptr(), &mut out)
        })?;
        Ok((out.re, out.im))
    }

    /// `<psi | P_q | psi>` for a single-site Pauli on qubit `q`,
    /// identity elsewhere.  Convenience helper around
    /// [`expect_pauli`].  Returns the real part only (single-site
    /// Pauli expectations are real up to roundoff).
    pub fn expect_pauli_single(&self, q: u32, pauli: PauliCode) -> Result<f64> {
        let n = self.num_qubits() as usize;
        if (q as usize) >= n {
            return Err(QuantumError::InvalidQubit {
                index: q as usize,
                max: n,
            });
        }
        let mut p = vec![PauliCode::I as u8; n];
        p[q as usize] = pauli as u8;
        let (re, _im) = self.expect_pauli(&p)?;
        Ok(re)
    }

    /// `P(Z_q = +1)` marginal probability.
    pub fn prob_z(&self, q: u32) -> Result<f64> {
        let mut out: f64 = 0.0;
        check(unsafe { moonlab_ca_peps_prob_z(self.ptr, q, &mut out) })?;
        Ok(out)
    }
}

impl Clone for CaPeps {
    fn clone(&self) -> Self {
        let ptr = unsafe { moonlab_ca_peps_clone(self.ptr) };
        assert!(!ptr.is_null(), "moonlab_ca_peps_clone returned NULL");
        Self { ptr }
    }
}

impl Drop for CaPeps {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { moonlab_ca_peps_free(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

fn check(rc: ca_peps_error_t) -> Result<()> {
    if rc == 0 {
        Ok(())
    } else {
        Err(QuantumError::Ffi(format!("ca_peps rc={rc}")))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fresh_state_traces_to_unit_z() {
        // |0...0> has <Z_q> = 1 on every qubit.
        let state = CaPeps::new(2, 2, 4).unwrap();
        assert_eq!(state.num_qubits(), 4);
        assert_eq!(state.lx(), 2);
        assert_eq!(state.ly(), 2);
        for q in 0..4 {
            let z = state.expect_pauli_single(q, PauliCode::Z).unwrap();
            assert!((z - 1.0).abs() < 1e-10, "<Z_{q}> = {} on |0000>", z);
        }
    }

    #[test]
    fn hadamard_zeros_out_z_expectation() {
        // H|0> = |+>; <Z> = 0.
        let mut state = CaPeps::new(2, 2, 4).unwrap();
        state.h(0).unwrap();
        let z0 = state.expect_pauli_single(0, PauliCode::Z).unwrap();
        assert!(z0.abs() < 1e-10, "<Z_0> = {} on H|0>", z0);
        // Other qubits stay at <Z> = 1.
        let z1 = state.expect_pauli_single(1, PauliCode::Z).unwrap();
        assert!((z1 - 1.0).abs() < 1e-10);
    }

    #[test]
    fn cnot_pair_yields_perfect_zz_correlation() {
        // Bell pair on (0, 1) via H(0) + CNOT(0, 1).
        let mut state = CaPeps::new(2, 1, 4).unwrap();
        state.h(0).unwrap();
        state.cnot(0, 1).unwrap();
        let zz = state.expect_pauli(&[PauliCode::Z as u8, PauliCode::Z as u8]).unwrap();
        assert!((zz.0 - 1.0).abs() < 1e-10, "<ZZ> = {} on Bell pair", zz.0);
    }

    #[test]
    fn rejects_invalid_dimensions() {
        assert!(CaPeps::new(0, 3, 4).is_err());
        assert!(CaPeps::new(3, 0, 4).is_err());
        assert!(CaPeps::new(2, 2, 0).is_err());
    }

    #[test]
    fn norm_starts_at_unit() {
        let state = CaPeps::new(2, 2, 4).unwrap();
        assert!((state.norm() - 1.0).abs() < 1e-10);
    }

    #[test]
    fn clone_is_independent() {
        let mut a = CaPeps::new(2, 2, 4).unwrap();
        a.h(0).unwrap();
        let b = a.clone();
        // After H(0) again on `a`, `a` returns to |0...0> but `b`
        // remains in (H|0>) (x) |0...>.
        a.h(0).unwrap();
        let z_a = a.expect_pauli_single(0, PauliCode::Z).unwrap();
        let z_b = b.expect_pauli_single(0, PauliCode::Z).unwrap();
        assert!((z_a - 1.0).abs() < 1e-10, "a back to |0>: <Z_0> = {}", z_a);
        assert!(z_b.abs() < 1e-10, "b still on |+>: <Z_0> = {}", z_b);
    }
}
